### ISSUE TITLE
Introduce delegate (wrapper) superclasses

### ISSUE THREAD

**Issue by jankotek** [eclipse/eclipse-collections#273]
I noticed EC uses delegate pattern in many cases. I am working on #262 and I need to introduce yet another wrapper. I would like to externalize delegate (wrapper) functionality into separate classes. 

I would like to add three type of super classes for primitive and object collection

* immutable wrapper - `AbstractUnmodifiableXXCollection` would extend this. Method which would throw 'not readable' exception would not be implemented in this wrapper

* untouchable wrapper - extends immutable wrapper + mutable methods, but does not implement methods which would expose wrapped collection somehow. Untouchable from Multi Reader collection would implement this

* mutable wrapper - extends untouchable wrapper + methods which expose wrapped collection (iterators...). No use in EC, but user might find it usefull for testing etc..

I am happy to send patch. But want some discussion about naming convention, backward compatibility....

**Comment by motlin** [comment]
I completely agree and had started down this path but never finished it in time for a major release. I have some unfinished commits on some of my branches with names like:

* Reduce code duplication by extracting a common superclass AbstractUnmodifiableRichIterable above all unmodifiable wrappers. Use serialization proxies to stabilize serialized forms in the future (breaks serialization).
* Reduce code duplication by extracting a common superclass AbstractSynchronizedRichIterable above all synchronized wrappers. Use serialization proxies to stabilize serialized forms in the future (breaks serialization).

The current problem with compatibility is just that the wrappers are serializable classes that extend serializable classes, so their serialized form is sensitive to the inheritance hierarchy and is generally bloated.
